src/ucs/datastruct/usage_tracker.c

```diff
@@ -0,0 +1,295 @@
+        return UCS_ERR_NO_ELEM;
+    }
+
+    item     = &kh_value(&usage_tracker->hash, iter);
+    *score_p = ucs_usage_tracker_score(item);
+    return UCS_OK;
+}
+
+static ucs_usage_tracker_element_t *
+ucs_usage_tracker_get_min_active(ucs_usage_tracker_h usage_tracker)
+{
+    ucs_usage_tracker_element_t *min_item = NULL;
+    unsigned active_count                 = 0;
+    ucs_usage_tracker_element_t *item;
+    khint_t k;
+
+    for (k = kh_begin(&usage_tracker->hash); k != kh_end(&usage_tracker->hash);
+         ++k) {
+        if (!kh_exist(&usage_tracker->hash, k)) {
+            continue;
+        }
+
+        item = &kh_val(&usage_tracker->hash, k);
+
+        if (!item->active) {
+            continue;
+        }
+
+        active_count++;
+
+        if ((min_item == NULL) || (ucs_usage_tracker_score(item) <
+                                   ucs_usage_tracker_score(min_item))) {
+            min_item = item;
+        }
+    }
+
+    return (active_count == usage_tracker->active_capacity) ? min_item : NULL;
+}
+
+static ucs_usage_tracker_element_t *
+ucs_usage_tracker_pushpop_active(ucs_usage_tracker_h usage_tracker,
+                                 ucs_usage_tracker_element_t *elem)
+{
+    ucs_usage_tracker_element_t *min_item;
+
+    min_item     = ucs_usage_tracker_get_min_active(usage_tracker);
+    elem->active = 1;
+
+    if (min_item != NULL) {
+        min_item->active = 0;
+    }
+
+    return min_item;
+}
+
+ucs_status_t
+ucs_usage_tracker_remove(ucs_usage_tracker_h usage_tracker, void *key)
+{
+    khint_t iter;
+
+    iter = kh_get(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key);
+    if (iter == kh_end(&usage_tracker->hash)) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    kh_del(usage_tracker_hash, &usage_tracker->hash, iter);
+    return UCS_OK;
+}
+
+void *ucs_usage_tracker_push_rx(ucs_usage_tracker_h usage_tracker, void *key,
+                                size_t score)
+{
+    ucs_usage_tracker_element_t *element, *ejected;
+    element     = ucs_usage_tracker_put(usage_tracker, key);
+    element->rx = score;
+
+    ejected = NULL;
+    if (!element->active) {
+        ejected = ucs_usage_tracker_pushpop_active(usage_tracker, element);
+    }
+
+    return ejected;
+}
+
+static int
+ucs_usage_tracker_is_important(ucs_usage_tracker_h usage_tracker, size_t score)
+{
+    ucs_usage_tracker_element_t *min_item;
+
+    if (score <=
+        usage_tracker->active_thresh * usage_tracker->ticks_per_flush) {
+        return 0;
+    }
+
+    min_item = ucs_usage_tracker_get_min_active(usage_tracker);
+    if (min_item == NULL) {
+        return 1;
+    }
+
+    return (score - ucs_usage_tracker_score(min_item)) >
+           usage_tracker->eject_thresh;
+}
+
+static void ucs_usage_tracker_flush_tx(ucs_usage_tracker_h usage_tracker)
+{
+    khint_t k;
+    ucs_usage_tracker_element_t *elem;
+
+    for (k = kh_begin(&usage_tracker->hash); k != kh_end(&usage_tracker->hash);

```

Comment:
use foreach macro?

Reply:
they copy kh_val instead of passing an address to it.